x-pack/plugins/ml/public/application/components/model_snapshots/model_snapshots_table.tsx

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { FC, useEffect, useState } from 'react';
+import { i18n } from '@kbn/i18n';
+
+// @ts-ignore
+import { formatDate } from '@elastic/eui/lib/services/format';
+import { FormattedMessage } from '@kbn/i18n/react';
+
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiInMemoryTable,
+  EuiLoadingSpinner,
+  EuiOverlayMask,
+  EuiConfirmModal,
+  EuiBasicTableColumn,
+} from '@elastic/eui';
+
+import { checkPermission } from '../../capabilities/check_capabilities';
+import { EditModelSnapshotFlyout } from './edit_model_snapshot_flyout';
+import { RevertModelSnapshotFlyout } from './revert_model_snapshot_flyout';
+import { ml } from '../../services/ml_api_service';
+import { JOB_STATE, DATAFEED_STATE } from '../../../../common/constants/states';
+import {
+  ModelSnapshot,
+  CombinedJobWithStats,
+} from '../../../../common/types/anomaly_detection_jobs';
+
+const TIME_FORMAT = 'YYYY-MM-DD HH:mm:ss';
+
+interface Props {
+  job: CombinedJobWithStats;
+  refreshJobList: () => void;
+}
+
+enum COMBINED_JOB_STATE {
+  OPEN_AND_RUNNING,
+  OPEN_AND_STOPPED,
+  CLOSED,
+  UNKNOWN,
+}
+
+export const ModelSnapshotTable: FC<Props> = ({ job, refreshJobList }) => {
+  const canCreateJob = checkPermission('canCreateJob');
+  const canStartStopDatafeed = checkPermission('canStartStopDatafeed');
+
+  const [snapshots, setSnapshots] = useState<ModelSnapshot[]>([]);
+  const [snapshotsLoaded, setSnapshotsLoaded] = useState<boolean>(false);
+  const [editSnapshot, setEditSnapshot] = useState<ModelSnapshot | null>(null);
+  const [revertSnapshot, setRevertSnapshot] = useState<ModelSnapshot | null>(null);
+  const [closeJobModalVisible, setCloseJobModalVisible] = useState<ModelSnapshot | null>(null);
+  const [combinedJobState, setCombinedJobState] = useState<COMBINED_JOB_STATE | null>(null);
+
+  useEffect(() => {
+    loadModelSnapshots();
+  }, []);
+
+  async function loadModelSnapshots() {
+    const { model_snapshots: ms } = await ml.getModelSnapshots(job.job_id);
+    setSnapshots(ms);
+    setSnapshotsLoaded(true);
+  }
+
+  async function checkJobIsClosed(snapshot: ModelSnapshot) {
```

Comment:
consider using `useCallback`, same about the other functions in this component

Reply:
Changed in ac22c8ce86162fd9d4db4687850de4ae33873f4b